### PR TITLE
Add tests for WriteUpgradeCancel.

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

Testing some (still) untested functions.

closes: #XXXX

### Commit Message / Changelog Entry

```bash
chore: add tests for writeupgradecancel.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 1]
```diff
@@ -754,6 +754,158 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestChanUpgradeCancel() {
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/upgrade_test.go:757]: move this further up to align with position as found in `upgrade.go`.

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 2]
```diff
@@ -754,6 +754,158 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
...
+	suite.Require().True(ok)
+
+	suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.WriteUpgradeCancelChannel(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, errorReceipt.Sequence)
+
+	channel := path.EndpointA.GetChannel()
+
+	// Verify that channel has been restored to previous state
+	suite.Require().Equal(types.OPEN, channel.State)
+	suite.Require().Equal(types.NOTINFLUSH, channel.FlushStatus)
+	suite.Require().Equal(errorReceipt.Sequence, channel.UpgradeSequence)
```
Comment (charleenfei) [inline modules/core/04-channel/keeper/upgrade_test.go:898]: we can also check the channel version here

**Review by charleenfei — APPROVED** [review]
one more check on the test but otherwise lgtm!

File: modules/core/04-channel/keeper/upgrade_test.go (hunk 2 above)
Comment (DimitrisJim) [reply modules/core/04-channel/keeper/upgrade_test.go:899]: yea, nice idea!

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 3]
```diff
@@ -754,6 +754,159 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
...
+			}
+
+			suite.Require().NoError(path.EndpointB.ChanUpgradeTry())
+
+			suite.Require().NoError(path.EndpointA.UpdateClient())
+
+			upgradeErrorReceiptKey := host.ChannelUpgradeErrorKey(path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID)
+			errorReceiptProof, proofHeight = suite.chainB.QueryProof(upgradeErrorReceiptKey)
+
+			var ok bool
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:848]: does this need to be declared as var? is there a reason we can't use `errorReceipt, ok := `

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 4]
```diff
@@ -754,6 +754,159 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
...
+	// cause the upgrade to fail on chain b so an error receipt is written.
+	suite.chainB.GetSimApp().IBCMockModule.IBCApp.OnChanUpgradeTry = func(
+		ctx sdk.Context, portID, channelID string, order types.Order, connectionHops []string, counterpartyVersion string,
+	) (string, error) {
+		return "", fmt.Errorf("mock app callback failed")
+	}
+
+	suite.Require().NoError(path.EndpointB.ChanUpgradeTry())
+
+	suite.Require().NoError(path.EndpointA.UpdateClient())
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:886]: I think I normally lean towards separating the assertion out from the endpoint call. But for sure we have a mixed bag across the codebase.

I think I just find it the tiniest bit more readable as:

```go
err := path.EndpointA.UpdateClient()
suite.Require().NoError(err)
```

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 5]
```diff
@@ -754,6 +754,159 @@ func (suite *KeeperTestSuite) TestChanUpgradeOpen() {
...
+	suite.Require().Equal(mock.Version, channel.Version)
+	suite.Require().Equal(errorReceipt.Sequence, channel.UpgradeSequence)
+
+	// Assert that state stored for upgrade has been deleted
+	upgrade, found := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.GetUpgrade(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+	suite.Require().Equal(types.Upgrade{}, upgrade)
+	suite.Require().False(found)
+	lastPacketSequence, found := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.GetCounterpartyLastPacketSequence(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+	suite.Require().Equal(uint64(0), lastPacketSequence)
+	suite.Require().False(found)
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:907]: whitespacing again :D 

```suggestion
	upgrade, found := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.GetUpgrade(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
	suite.Require().Equal(types.Upgrade{}, upgrade)
	suite.Require().False(found)
	
	lastPacketSequence, found := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.GetCounterpartyLastPacketSequence(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
	suite.Require().Equal(uint64(0), lastPacketSequence)
	suite.Require().False(found)
```

**Review by damiannolan — APPROVED** [review]
Great work, again just some nits. LGTM! :)

File: modules/core/04-channel/keeper/upgrade_test.go (hunk 1 above)
Comment (DimitrisJim) [reply modules/core/04-channel/keeper/upgrade_test.go:757]: yeah yeah, can fix the extra nits tho

File: modules/core/04-channel/keeper/upgrade_test.go (hunk 3 above)
Comment (DimitrisJim) [reply modules/core/04-channel/keeper/upgrade_test.go:849]: yea, rebinding messes up our modifications in the malleate

Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade_test.go:849]: aha, yes! I see :)

**Pull request merged**